File: Toán UD/CK/cpp/perfect_num.cpp
#include <iostream>
#include <vector>
using namespace std;

bool checkPerfect(int n) {
    int sum = 0;
    for (int i = 1; i < n; ++i) {
        if (n % i == 0) {
            sum += i;
        }
    }
    return sum == n;
}

vector<int> findPerfectNumbers(int n) {
    vector<int> result;
    for (int i = 1; i < n; ++i) {
        if (checkPerfect(i)) {
            result.push_back(i);
        }
    }
    return result;
}

int main() {
    int n;
    cout << "n = ";
    cin >> n;

    vector<int> perfectNumbersList = findPerfectNumbers(n);

    cout << "Perfect numbers from 1 to " << n << ": ";
    for (size_t i = 0; i < perfectNumbersList.size(); ++i) {
        cout << perfectNumbersList[i] << " ";
    }
    cout << endl;

    return 0;
}

File: Toán UD/CK/ex/perfect_num.py
# Bài tập 2: Tìm số hoàn hảo từ 1 đến n.

def check_perfect(n):
    sum = 0
    for i in range(1, n):
        if (n % i) == 0:
            sum += i
    if sum == n:
        return True
    else:
        return False

def find_perfect_numbers(n):
    rs = []
    for i in range(1, n):
        if check_perfect(i):
            rs.append(i)
    return rs

n = int(input("n = "))

perfect_numbers_list = find_perfect_numbers(n)

print(perfect_numbers_list)